src/nix/flake.cc

```diff
@@ -1126,267 +754,158 @@ struct CmdFlakeShow : FlakeCommand, MixJSON
-                        || attrPathS[0] == "overlays"
-                        )) {
-                    for (const auto &subAttr : visitor2->getAttrs()) {
-                        if (hasContent(*visitor2, attrPath2, subAttr)) {
-                            return true;
-                        }
-                    }
-                    return false;
-                }
-
-                // If we don't recognize it, it's probably content
-                return true;
-            } catch (EvalError & e) {
-                // Some attrs may contain errors, e.g. legacyPackages of
-                // nixpkgs. We still want to recurse into it, instead of
-                // skipping it at all.
-                return true;
-            }
-        };
+        auto [cache, inventory] = flake_schemas::call(*state, flake, getDefaultFlakeSchemas());
 
-        std::function<nlohmann::json(
-            eval_cache::AttrCursor & visitor,
-            const std::vector<Symbol> & attrPath,
-            const std::string & headerPrefix,
-            const std::string & nextPrefix)> visit;
-
-        visit = [&](
-            eval_cache::AttrCursor & visitor,
-            const std::vector<Symbol> & attrPath,
-            const std::string & headerPrefix,
-            const std::string & nextPrefix)
-            -> nlohmann::json
-        {
-            auto j = nlohmann::json::object();
-
-            auto attrPathS = state->symbols.resolve(attrPath);
-
-            Activity act(*logger, lvlInfo, actUnknown,
-                fmt("evaluating '%s'", concatStringsSep(".", attrPathS)));
+        if (json) {
+            std::function<void(ref<eval_cache::AttrCursor> node, nlohmann::json & obj)> visit;
 
-            try {
-                auto recurse = [&]()
-                {
-                    if (!json)
-                        logger->cout("%s", headerPrefix);
-                    std::vector<Symbol> attrs;
-                    for (const auto &attr : visitor.getAttrs()) {
-                        if (hasContent(visitor, attrPath, attr))
-                            attrs.push_back(attr);
-                    }
+            visit = [&](ref<eval_cache::AttrCursor> node, nlohmann::json & obj)
+            {
+                flake_schemas::visit(
+                    showAllSystems ? std::optional<std::string>() : localSystem,
+                    node,
+
+                    [&](ref<eval_cache::AttrCursor> leaf)
+                    {
+                        obj.emplace("leaf", true);
+
+                        if (auto what = flake_schemas::what(leaf))
+                            obj.emplace("what", what);
+
+                        if (auto shortDescription = flake_schemas::shortDescription(leaf))
+                            obj.emplace("shortDescription", shortDescription);
+
+                        if (auto drv = flake_schemas::derivation(leaf))
+                            obj.emplace("derivationName", drv->getAttr(state->sName)->getString());
```

Comment:
Apart from recognising `what`, `shortDescription` and `derivation` as the `leaf` types, it would be nice to also allow setting arbitrary values. 

Perhaps like:
```patch
From 1d23c1e871981f5666a12c4409bd0574fc1e1e02 Mon Sep 17 00:00:00 2001
From: shivaraj-bh <sbh69840@gmail.com>
Date: Wed, 21 Aug 2024 15:14:03 +0530
Subject: [PATCH] support value as leaf node

---
 src/nix/flake.cc | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/nix/flake.cc b/src/nix/flake.cc
index f470d847bf8..6960a3e9e89 100644
--- a/src/nix/flake.cc
+++ b/src/nix/flake.cc
@@ -18,6 +18,7 @@
 #include "markdown.hh"
 #include "users.hh"
 #include "flake-schemas.hh"
+#include "value-to-json.hh"
 
 #include <nlohmann/json.hpp>
 #include <queue>
@@ -782,6 +783,13 @@ struct CmdFlakeShow : FlakeCommand, MixJSON, flake_schemas::MixFlakeSchemas
                         if (auto drv = flake_schemas::derivation(leaf))
                             obj.emplace("derivationName", drv->getAttr(state->sName)->getString());
 
+                        // TODO: Add a function in flake-schemas.hh to handle this
+                        if (auto value = leaf->maybeGetAttr("value")) {
+                            auto v = value->forceValue();
+                            NixStringContext context;
+                            obj.emplace("value", printValueAsJSON(*state, true, v, noPos, context, false));
+                        }
+
                         // FIXME: add more stuff
                     },
 
```

If this is out of scope of this PR, I could raise a separate PR after this is merged, describing the use-case in more detail.